ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,428 @@
+deleted, the values can be Retain or Delete.
+
+**Immutable object**: yes
+
+**Scope**: cluster
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTargetClass
+metadata:
+    name: isolation-level-namespace
+spec:
+    provisioner: appstudio.redhat.com/devsandbox
+    parameters: {}
+    reclaimPolicy: Delete
+```
+
+#### Environment
+
+[Environment] objects refer to a [DTC] using the `deploymentTargetClaim`
+field. The environment controller will wait for the [DTC] to get to the
+“bound” phase, once it is bound, it will reach the [DT] and read the
+target's connection details from the kubernetesCredentials field and
+configure Argo/Gitops services to use them.
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: Environment
+metadata:
+    name: prod
+spec:
+    displayName: “Production for Team A”
+    deploymentStrategy: AppStudioAutomated
+    parentEnvironment: staging
+    tags: [prod]
+    configuration:
+        target:
+            deploymentTargetClaim:
+                claimName: prod-dtc
+```
+
+### Controllers
+
+#### DeploymentTargetBinder
+
+Binds [DeploymentTargetClaim] to a [DeploymentTarget] that satisfies its requirements.
+
+It watches for the creation of new [DTC] and tries to find a matching [DT] for each one of them.
+
+A [DT] that we created dynamically for a specific [DTC] will always be attached to it.
+
+[DT] and [DTC] have one to one bindings.
+
+##### Binding Loop (DTC Reconciliation)
+
+![](../diagrams/ADR-0008/binding-controller.jpg)
+
+#### DeploymentTargetProvisioner
+
+Watch for the creation of a new [DTC]. If the [DTCLS] of the [DTC] matches a [DTCLS] the provisioner was
+configured for, it reads the parameters from that class, provisions the target and creates a [DT]
+object which references the [DTC] that started the process.
+
+When a [DTC] is deleted, if it was bound to a [DT] created by the provisioner, it reclaims the [DT] and
+the actual cluster that was created for it based on the reclaimPolicy configuration.
+
+##### Provision Loop (DTC Reconciliation)
+
+![](../diagrams/ADR-0008/provision-loop.jpg)
+
+##### Deprovision Loop (DT Reconciliation)
+
+![](../diagrams/ADR-0008/deprovision-loop.jpg)
+
+#### EnvironmentController
+
+TODO - anything to write about the Environment controller?
+
+### DeploymentTarget(DT) and DeploymentTargetClaim(DTC) Lifecycle
+
+![](../diagrams/ADR-0008/dt-dtc-lifecycle.jpg)
+
+### Access Matrix
+
+| *User,Controller/CRD* | **DT**                | **DTC**               | **DTCLS** | **Environment**      |
+|-----------------------|-----------------------|-----------------------|-----------|----------------------|
+| **Binder**            | watch/list/get/update | watch/list/get/update |           |                      |
+| **Provionser**        | create/delete         | watch                 | get       |                      |
```

Comment:
What has the ability to create/delete/update DTCLS?

Reply:
typo: Provisioner

Reply:
Only we (appstudio administrators) would create new DTCLS resources when we want to make a new kind of deployment target available for automatic provisioning.

Reply:
> typo: Provisioner

Fixed.